component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -361,24 +366,52 @@ private boolean isValidSignature(String clientId, SignedJWT signedJWT, String te
                                      String alias) throws OAuthClientAuthnException {
 
         X509Certificate cert = null;
+        String jwksUri = "";
+        boolean isValidSignature = false;
         try {
             cert = (X509Certificate) OAuth2Util.getX509CertOfOAuthApp(clientId, tenantDomain);
         } catch (IdentityOAuth2Exception e) {
             if (log.isDebugEnabled()) {
                 log.debug(e.getMessage(), e);
             }
         }
+        //if cert is null check whether a jwks endpoint is configured for the service provider
+        if (cert == null) {
+            try {
+                ServiceProviderProperty[] spProperties = OAuth2Util.getServiceProvider(clientId).getSpProperties();
+                for (ServiceProviderProperty spProperty : spProperties) {
+                    if (Constants.JWKS_URI.equals(spProperty.getName())) {
+                        jwksUri = spProperty.getValue();
+                    }
+                }
+                //validate the signature of the assertion using the jwks end point
+                String jwtString = signedJWT.getParsedString();
+                String alg = signedJWT.getHeader().getAlgorithm().getName();
+                Map<String, Object> options = new HashMap<String, Object>();
+                isValidSignature = new JWKSBasedJWTValidator().validateSignature(jwtString, jwksUri, alg, options);
+
+            } catch (IdentityOAuth2Exception e) {
+                throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
+            }
+        }
+
         // If certificate is not configured in service provider, it will throw an error.
         // For the existing clients need to handle that error and get from truststore.
-        if (cert == null) {
+        if (StringUtils.isBlank(jwksUri) && cert == null) {
+
             cert = getCertificate(tenantDomain, alias);
         }
 
-        try {
-            return validateSignature(signedJWT, cert);
-        } catch (JOSEException e) {
-            throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
+        if (StringUtils.isBlank(jwksUri) && cert != null) {
+            try {
+                isValidSignature = validateSignature(signedJWT, cert);
+            } catch (JOSEException e) {
+                throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
```

Comment:
Use the constructor with throwable if exists.

Reply:
done